README.rst
=========================
Discourse SSO application
=========================

| |travis| |coveralls| |landscape| |scrutinizer| |codacy| |license|

.. |travis| image:: http://img.shields.io/travis/fmarco76/DiscourseSSO/master.png?style=flat
    :alt: Travis-CI Build Status
    :target: https://travis-ci.org/fmarco76/DiscourseSSO

.. |coveralls| image:: https://coveralls.io/repos/fmarco76/DiscourseSSO/badge.svg?style=flat
    :alt: Coverage Status
    :target: https://coveralls.io/r/fmarco76/DiscourseSSO

.. |landscape| image:: https://landscape.io/github/fmarco76/DiscourseSSO/master/landscape.svg?style=flat
    :target: https://landscape.io/github/fmarco76/DiscourseSSO/master
    :alt: Code Quality Status

.. |scrutinizer| image:: https://img.shields.io/scrutinizer/g/fmarco76/DiscourseSSO/master.png?style=flat
    :alt: Scrutinizer Status
    :target: https://scrutinizer-ci.com/g/fmarco76/DiscourseSSO/

.. |codacy| image:: https://www.codacy.com/project/badge/ff8b39c8455d4f2ca4ead7e034f6b6d6
    :alt: Codacy Status
    :target: https://www.codacy.com/app/marco-fargetta/DiscourseSSO

.. |license| image:: https://img.shields.io/github/license/fmarco76/DiscourseSSO.svg?style=flat 
    :alt: License
    :target: http://www.apache.org/licenses/LICENSE-2.0.txt


Discourse SSO application implements the SSO protocol requested by `discourse
forum application <http://www.discourse.org>`_. The real authentication is
performed by the web server running the application which is responsible of
verify the requests coming from discourse and prepare the token to send back
after the authentication.


Requirements
------------

DiscourseSSO application can execute in any web server supporting python code.
However it has been tested only with *apache httpd server* using *mod_wsgi* and
these are described in the installation section.

The authentication has to be performed by the web server and this require an additional
module. The module has to provide the user information the in application environment.
This is the case with many modules like *mod_shibboleth* for **SAML** described
in the installation.


Installation
------------

For the installation we consider Discourse container has been installed in the server
and work properly with local accounts.

Prepare the server where DiscourseSSO has to be deployed. The server could be
the same of the Discourse server using the same *nginx* to execute the sso code or an
additional *apache httpd* server taking care of not create conflict with the ports.

The installation of httpd server has to include *mod_wsgi* to execute the code, *mod_ssl*
to have the authentication over https and *mod_shib* (or equivalent) to implement the
authentication.

After the installation of all modules httpd has to be configured to work on the ssl port.
If mod_ssl comes with your distribution it should include the basic configuration,
otherwise have a look at the `apache httpd documentation <http://httpd.apache.org/docs/>`_
for your specific version. The standard http port **80** will not be used so it could be
disabled.

The next step is to configure mod_shib in order to allow the authentication
using saml (for details on mod_shib configuration look at the `official wiki
<https://wiki.shibboleth.net/confluence/display/SHIB2/NativeSPConfiguration>`_).

When the server is properly configured download/clone the **DiscourseSSO** package
from `GitHub <https://github.com/fmarco76/DiscourseSSO>`_ in a directory accessible
by apache. Configure mod_wsgi to include the source code of the package and add
the ``DiscourseSSO.wsgi`` script to the location to use for Discourse. Finally, shibboleth
has to protect the authentication directory so the user has to go to the IdP in order
be accepted. The configuration should look like: ::

    WSGIDaemonProcess discourse threads=5 python-path=<path-to-discourse>/DiscourseSSO/src
    WSGIScriptAlias /DiscourseSSO /var/www/DiscourseSSO/DiscourseSSO.wsgi
    WSGISocketPrefix /var/run/wsgi
    .
    .
    <Directory /var/www/DiscourseSSO>
        WSGIProcessGroup discourse
        WSGIApplicationGroup %{GLOBAL}
        Order deny,allow
        Allow from all
    </Directory>
    <Location /DiscourseSSO/sso/auth>
        AuthType shibboleth
        ShibCompatWith24 On
        ShibRequestSetting requireSession 1
        require shib-session
    </Location>

Finally, both Discourse and DiscourseSSO need to be configured. Enable the sso in
Discourse following the `official documentation <https://meta.discourse.org/t/official-single-sign-on-for-discourse/13045>`_.
The sso url to is the one going to your DiscourseSSO installation plus ``sso/login``, so using
the above configuration the url is ``https://<your-domain>/DisocurseSSO/sso/login``.

The sso_secret is a random string and has to be the same in both services. The configuration
file of DiscourseSSO is ``src/discoruseSSO/config.py`` and it require the secret key (``DISCOURSE_SECRET_KEY``),
the url of Discourse (``DISCOURSE_URL``) and the name of the environment variables
where mod_shib will provide the user information(``DISCOURSE_USER_MAP``). This is a
map where the key is the attribute provided back to Discourse whereas the values
are the name of the variables to lookup. The name can be generated combining different
values but the other accept only one value. Default are good for a SAML based
authentication but for other authentication mechanism you have to modify them accordingly.

After the configuration restart the apache httpd daemon and enable the sso in Discourse.


CHANGELOG.rst
07 June 2017: Extend the map with all the possible parameter and add a
              new list for the other using filters

tests/test_sso.py
# Copyright 2015 INFN
# All Rights Reserved.
#
#    Licensed under the Apache License, Version 2.0 (the "License"); you may
#    not use this file except in compliance with the License. You may obtain
#    a copy of the License at
#
#         http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
#    License for the specific language governing permissions and limitations
#    under the License.

"""
SSO Application tests
"""


from flask import url_for
import pytest
from urlparse import urlparse
from werkzeug.exceptions import BadRequest, Forbidden
from discourseSSO import sso

app = sso.app


class Test_sso():

    def test_payload_check(self):
        """Test the payload is properly managed and the user is sent to the
        authentication page
        """
        with app.test_request_context('/sso/login?sso=bm9uY2U9Y2I2ODI1MWVlZm'
                                      'I1MjExZTU4YzAwZmYxMzk1ZjBjMGI%3D%0A&'
                                      'sig=2828aa29899722b35a2f191d34ef9b3ce'
                                      '695e0e6eeec47deb46d588d70c7cb56',
                                      method='GET'):
            res = sso.payload_check()
            assert res.status_code == 302
            assert urlparse(res.location).path == url_for('user_authz')

    def test_bad_payload_sig(self):
        """Test the error code 400 is sent if the signature do not match
        the payload
        """
        with app.test_request_context('/sso/login?sso=bm9uY2U9Y2I2ODI1MWVlZm'
                                      'I1MjExZTU4YzAwZmYxMzk1ZjBjMGI%3D%0A&'
                                      'sig=2828aa29899722b35a2f191d34ef9b3ce'
                                      '695e0e6eeec47deb46d588d70c7cb58',
                                      method='GET'):
            with pytest.raises(BadRequest):
                sso.payload_check()

    def test_no_payload(self):
        """Test the error code 400 is sent if the sso field is not provided"""
        with app.test_request_context('/sso/login?sig=2828aa29899722b35a2f191'
                                      'd34ef9b3ce695e0e6eeec47deb46d588d70c7c'
                                      'b56',
                                      method='GET'):
            with pytest.raises(BadRequest):
                sso.payload_check()

    def test_no_hash(self):
        """Test the error code 400 is sent if the sig field is not provided"""
        with app.test_request_context('/sso/login?sso=bm9uY2U9Y2I2ODI1MWVlZm'
                                      'I1MjExZTU4YzAwZmYxMzk1ZjBjMGI%3D%0A&',
                                      method='GET'):
            with pytest.raises(BadRequest):
                sso.payload_check()

    def test_authentication_no_shibboleth_attributes(self):
        """Test the authentication when shibboleth do not provide attributes"""
        with app.test_request_context('/sso/auth',
                                      method='GET'):
            with pytest.raises(Forbidden):
                sso.user_authz()

    def test_authentication_no_previous_session(self):
        """Test the authentication are properly send to Discourse"""
        with app.test_request_context('/sso/auth',
                                      method='GET',
                                      environ_base={
                                          'givenName': 'sam',
                                          'sn': '',
                                          'username': 'samsam',
                                          'mail': 'test@test.com',
                                          'eppn': 'hello123'}
                                      ):
            with pytest.raises(Forbidden):
                sso.user_authz()

    def test_authentication_generation(self):
        """Test the authentication are properly send to Discourse"""
        with app.test_request_context('/sso/auth',
                                      method='GET',
                                      environ_base={
                                          'givenName': 'sam',
                                          'sn': '',
                                          'username': 'samsam',
                                          'mail': 'test@test.com',
                                          'eppn': 'hello123'}
                                      ) as req:
            req.session['nonce'] = 'nonce=cb68251eefb5211e58c00ff1395f0c0b'
            resp = sso.user_authz()
            assert resp.status_code == 302
            # sso and sig are different from the one reported in
            # https://meta.discourse.org/t/official-single-sign-on-for-
            #        discourse/13045
            # This because ruby and python include new lines in different
            # positions during the base64 encoding (of course they do not
            # matter for the base64 but the following URLencoding and
            # signature are slightly different)
            assert resp.location == ('http://discuss.example.com/session/'
                                     'sso_login?sso=bm9uY2U9Y2I2ODI1MWVlZ'
                                     'mI1MjExZTU4YzAwZmYxMzk1ZjBjMGImbmFt'
                                     'ZT1zYW0mdXNlcm5hbWU9%0Ac2Ftc2FtJmVt'
                                     'YWlsPXRlc3QlNDB0ZXN0LmNvbSZleHRlcm5'
                                     'hbF9pZD1oZWxsbzEyMyZhZG1pbj1m%0AYWx'
                                     'zZQ%3D%3D%0A&sig=a8ad52d665ddf2d2d5'
                                     '5de5d08d745f46d44a503d0b51b0273dd95'
                                     'e1f2abe1cbd')

    def test_authentication_generation_with_full_name(self):
        """Test the authentication are properly send to Discourse"""
        with app.test_request_context('/sso/auth',
                                      method='GET',
                                      environ_base={
                                          'givenName': 'sam',
                                          'sn': 'big',
                                          'mail': 'test@test.com',
                                          'eppn': 'hello123'}
                                      ) as req:
            req.session['nonce'] = 'nonce=cb68251eefb5211e58c00ff1395f0c0b'
            resp = sso.user_authz()
            assert resp.status_code == 302
            assert resp.location == ('http://discuss.example.com/session/'
                                     'sso_login?sso=bm9uY2U9Y2I2ODI1MWVlZ'
                                     'mI1MjExZTU4YzAwZmYxMzk1ZjBjMGImbmFt'
                                     'ZT1zYW0gYmlnJnVzZXJu%0AYW1lPXNhbWJp'
                                     'Z19iNjQyJmVtYWlsPXRlc3QlNDB0ZXN0LmN'
                                     'vbSZleHRlcm5hbF9pZD1oZWxsbzEy%0AMyZ'
                                     'hZG1pbj1mYWxzZQ%3D%3D%0A&sig=8177ae'
                                     '45c294212a96767cfb2208db867a14fa099'
                                     '0bf7efb2f36dcac41d563e8')


    def test_authentication_generation_with_avatar_bio(self):
        """Test the authentication are properly send to Discourse"""
        with app.test_request_context('/sso/auth',
                                      method='GET',
                                      environ_base={
                                          'givenName': 'sam',
                                          'sn': '',
                                          'username': 'samsam',
                                          'mail': 'test@test.com',
                                          'eppn': 'hello123',
                                          'avatar': 'http://myAvatarURL',
                                          'profile': 'http://myProfileURL'}
                                      ) as req:
            req.session['nonce'] = 'nonce=cb68251eefb5211e58c00ff1395f0c0b'
            resp = sso.user_authz()
            assert resp.status_code == 302
            # sso and sig are different from the one reported in
            # https://meta.discourse.org/t/official-single-sign-on-for-
            #        discourse/13045
            # This because ruby and python include new lines in different
            # positions during the base64 encoding (of course they do not
            # matter for the base64 but the following URLencoding and
            # signature are slightly different)
            assert resp.location == ('http://discuss.example.com/session/'
                                     'sso_login?sso=bm9uY2U9Y2I2ODI1MWVlZ'
                                     'mI1MjExZTU4YzAwZmYxMzk1ZjBjMGImbmFt'
                                     'ZT1zYW0mdXNlcm5hbWU9%0Ac2Ftc2FtJmVt'
                                     'YWlsPXRlc3QlNDB0ZXN0LmNvbSZleHRlcm5'
                                     'hbF9pZD1oZWxsbzEyMyZhdmF0YXJf%0AdXJ'
                                     'sPWh0dHAlM0EvL215QXZhdGFyVVJMJmJpbz'
                                     '1odHRwJTNBLy9teVByb2ZpbGVVUkwmYWRta'
                                     'W49%0AZmFsc2U%3D%0A&sig=61504842b6a'
                                     '130d0f2d6976de814313a8df539d5e95bd9'
                                     '32d693acbcf0b9df14')

    def test_authentication_generation_with_flags(self):
        """Test the authentication are properly send to Discourse"""
        with app.test_request_context('/sso/auth',
                                      method='GET',
                                      environ_base={
                                          'givenName': 'sam',
                                          'sn': '',
                                          'username': 'samsam',
                                          'mail': 'test@test.com',
                                          'eppn': 'my.name@my.idp',
                                          'avatar': 'http://myAvatarURL',
                                          'profile': 'http://myProfileURL'}
                                      ) as req:
            req.session['nonce'] = 'nonce=cb68251eefb5211e58c00ff1395f0c0b'
            resp = sso.user_authz()
            assert resp.status_code == 302
            # sso and sig are different from the one reported in
            # https://meta.discourse.org/t/official-single-sign-on-for-
            #        discourse/13045
            # This because ruby and python include new lines in different
            # positions during the base64 encoding (of course they do not
            # matter for the base64 but the following URLencoding and
            # signature are slightly different)
            assert resp.location == ('http://discuss.example.com/session/'
                                     'sso_login?sso=bm9uY2U9Y2I2ODI1MWVlZ'
                                     'mI1MjExZTU4YzAwZmYxMzk1ZjBjMGImbmFt'
                                     'ZT1zYW0mdXNlcm5hbWU9%0Ac2Ftc2FtJmVt'
                                     'YWlsPXRlc3QlNDB0ZXN0LmNvbSZleHRlcm5'
                                     'hbF9pZD1teS5uYW1lJTQwbXkuaWRw%0AJmF'
                                     '2YXRhcl91cmw9aHR0cCUzQS8vbXlBdmF0YX'
                                     'JVUkwmYmlvPWh0dHAlM0EvL215UHJvZmlsZ'
                                     'VVS%0ATCZhZG1pbj1mYWxzZSZyZXF1aXJlX'
                                     '2FjdGl2YXRpb249ZmFsc2U%3D%0A&sig=26'
                                     '8beaa221824d9c5ec9df3cb85e0655e86e1'
                                     'ba49ce516155f3f2557d7340140')

    def test_error_page_403(self):
        """Test the correct error code is propagated"""
        with app.test_request_context('/sso/auth',
                                      method='GET',
                                      environ_base={
                                          'givenName': 'sam',
                                          'sn': '',
                                          'username': 'samsam',
                                          'mail': 'test@test.com',
                                          'eppn': 'hello123'}
                                      ):
            resp = sso.attribuete_not_provided(None)
            assert resp[1] == 403

src/discourseSSO/config.py
# Copyright 2015 INFN
# All Rights Reserved.
#
#    Licensed under the Apache License, Version 2.0 (the "License"); you may
#    not use this file except in compliance with the License. You may obtain
#    a copy of the License at
#
#         http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
#    License for the specific language governing permissions and limitations
#    under the License.

"""
SSO FLASK Application for Discourse main configuration file
"""

# Discourse URL to send the user back
DISCOURSE_URL = 'http://discuss.example.com'

# Secret key shared with the Discourse server
DISCOURSE_SECRET_KEY = 'd836444a9e4084d5b224a60c208dce14'

# Attribute to read from the environment after user validation
DISCOURSE_USER_MAP = {
    'name': ['givenName', 'sn'],
    'username': 'username',
    'external_id': 'eppn',
    'email': 'mail',
    'avatar_url': 'avatar',
    'bio': 'profile',
}

# Flags associated with created accounts. Possible names are:
# require_activation, admin, moderator and suppress_welcome_message.
# If the same name is repeated it is applied multiple times in the
# provided order and last accepted value will be used.
# The filter is analysed and if it match then the value is added to the
# user information sent to discourse. The filter format is key=value where
# the key is an environment attribute and the value is matched using python
# regular expression. If filter is not provided the flag is always provided
DISCOURSE_USER_FLAGS = [
    {
        'name': 'require_activation',
        'value': 'false',
        'filter': 'eppn=^my.name@my.idp$',
    },
    {
        'name': 'admin',
        'value': 'false',
    },
]


src/discourseSSO/default.py
# Copyright 2015 INFN
# All Rights Reserved.
#
#    Licensed under the Apache License, Version 2.0 (the "License"); you may
#    not use this file except in compliance with the License. You may obtain
#    a copy of the License at
#
#         http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
#    License for the specific language governing permissions and limitations
#    under the License.

"""
Default configuration for FLASK app
"""


class Config(object):
    DEBUG = False
    TESTING = False
    SECRET_KEY = 'vWr,-n7NlGPv9SyIGBMr4ehwThUY92DpWPqIuh2NP_6Of-_8b3,h'
    # Discourse URL to send the user back
    DISCOURSE_URL = 'http://discuss.example.com'
    # Secret key shared with the Discourse server
    DISCOURSE_SECRET_KEY = 'd836444a9e4084d5b224a60c208dce14'
    # Attribute to read from the environment after user validation
    DISCOURSE_USER_MAP = {
        'name': ['givenName', 'sn'],
        'username': 'username',
        'external_id': 'eppn',
        'email': 'mail'
    }


src/discourseSSO/sso.py
# Copyright 2015 INFN
# All Rights Reserved.
#
#    Licensed under the Apache License, Version 2.0 (the "License"); you may
#    not use this file except in compliance with the License. You may obtain
#    a copy of the License at
#
#         http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
#    License for the specific language governing permissions and limitations
#    under the License.

"""
SSO FLASK Application for Discourse
The configuration file is defined with the variable "DISCOURSE_SSO_CONFIG",
for the most significant values look at the sso/default.py file
"""


from flask import abort, Flask, redirect, render_template, request, url_for, \
    session

import base64
import hashlib
import hmac
import urllib
import re

app = Flask(__name__)
app.config.from_object('discourseSSO.default.Config')
app.config.from_pyfile('config.py')
app.config.from_envvar('DISCOURSE_SSO_CONFIG', True)


@app.route('/sso/login')
def payload_check():
    """
    Verify the payload and signature coming from a Discourse server and if
    correct redirect to the authentication page
    :return: The redirection page to the authentication page
    """
    payload = request.args.get('sso', '')
    signature = request.args.get('sig', '')

    app.logger.debug('Request to login with payload="%s" signature="%s"',
                     payload, signature)
    if not payload or not signature:
        abort(400)

    app.logger.debug('Session Secret Key: %s',
                     app.secret_key)
    app.logger.debug('SSO Secret Key: %s',
                     app.config.get('DISCOURSE_SECRET_KEY'))
    dig = hmac.new(
        app.config.get('DISCOURSE_SECRET_KEY'),
        payload,
        hashlib.sha256
    ).hexdigest()
    app.logger.debug('Calculated hash: %s', dig)
    if dig != signature:
        abort(400)
    decoded_msg = base64.decodestring(payload)
    session['nonce'] = decoded_msg
    return redirect(url_for('user_authz'))


@app.route('/sso/auth')
def user_authz():
    """
    Read the user attributes provided by the application server (generally
    it is apache httpd) as environment variables and create the payload to
    send to discourse
    :return: The redirection page to Discourse
    """
    attribute_map = app.config.get('DISCOURSE_USER_MAP')
    user_flag_filters = app.config.get('DISCOURSE_USER_FLAGS')
    email = request.environ.get(attribute_map['email'])
    external_id = request.environ.get(attribute_map['external_id'])
    if not (email and external_id):
        abort(403)
    name_list = []
    for name_to_map in attribute_map['name']:
        if request.environ.get(name_to_map):
            name_list.append(request.environ.get(name_to_map))
    name = ' '.join(name_list)
    if request.environ.get(attribute_map['username']):
        username = request.environ.get(attribute_map['username'])
    else:
        username = (name.replace(' ', '') +
                    "_" +
                    hashlib.md5(email).hexdigest()[0:4]
                    )
    avatar_url = request.environ.get(attribute_map['avatar_url'])
    bio = request.environ.get(attribute_map['bio'])
    app.logger.debug('Authenticating "%s" with username "%s" and email "%s"',
                     name, username, email)
    if 'nonce' not in session:
        abort(403)
    query = (session['nonce'] +
             '&name=' + name +
             '&username=' + username +
             '&email=' + urllib.quote(email) +
             '&external_id=' + urllib.quote(external_id))
    if avatar_url:
        query = query + '&avatar_url=' + urllib.quote(avatar_url)
    if bio:
        query = query + '&bio=' + urllib.quote(bio)
    flags = {}
    for user_flag in user_flag_filters:
        if 'filter' in user_flag:
            filter = user_flag['filter'].split('=')
            reg_exp = re.compile(filter[1])
            if (request.environ.get(filter[0]) and
                    reg_exp.match(request.environ.get(filter[0]))):
                flags[user_flag['name']] = user_flag['value']
        else:
            flags[user_flag['name']] = user_flag['value']
    for flags_name in sorted(flags.keys()):
        query = query + '&' + flags_name + '=' + flags[flags_name]
    app.logger.debug('Query string to return: %s', query)
    query_b64 = base64.encodestring(query)
    app.logger.debug('Base64 query string to return: %s', query_b64)
    query_urlenc = urllib.quote(query_b64)
    app.logger.debug('URLEnc query string to return: %s', query_urlenc)
    sig = hmac.new(
        app.config.get('DISCOURSE_SECRET_KEY'),
        query_b64,
        hashlib.sha256
    ).hexdigest()
    app.logger.debug('Signature: %s', sig)
    redirect_url = (app.config.get('DISCOURSE_URL') +
                    '/session/sso_login?'
                    'sso=' + query_urlenc +
                    '&sig=' + sig)

    return redirect(redirect_url)


@app.errorhandler(403)
def attribuete_not_provided(error):
    """
    Render a custom error page in case the IdP authenticate the user but does
    not provide the requested attributes

    :type error: object
    """
    return render_template('403.html'), 403
